src/themis/secure_cell_seal_passphrase.c

```diff
@@ -0,0 +1,341 @@
+#define DEFAULT_AUTH_TOKEN_SIZE                                                                                \
+    (themis_scell_auth_token_passphrase_min_size + THEMIS_AUTH_SYM_IV_LENGTH + THEMIS_AUTH_SYM_AUTH_TAG_LENGTH \
+     + themis_scell_pbkdf2_context_min_size + THEMIS_AUTH_SYM_PBKDF2_SALT_LENGTH)
+
+static inline uint32_t soter_alg_without_kdf(uint32_t alg)
+{
+    /* Explicit cast for mask to have correct size before negation */
+    return (alg & ~((uint32_t)SOTER_SYM_KDF_MASK)) | SOTER_SYM_NOKDF;
+}
+
+static inline size_t soter_alg_key_length(uint32_t alg)
+{
+    return (alg & SOTER_SYM_KEY_LENGTH_MASK) / 8;
+}
+
+static inline bool soter_alg_reserved_bits_valid(uint32_t alg)
+{
+    static const uint32_t used_bits = SOTER_SYM_KEY_LENGTH_MASK | SOTER_SYM_PADDING_MASK
+                                      | SOTER_SYM_KDF_MASK | SOTER_SYM_ALG_MASK;
+    return (alg & ~used_bits) == 0;
+}
+
+themis_status_t themis_auth_sym_encrypt_message_with_passphrase_(const uint8_t* passphrase,
+                                                                 size_t passphrase_length,
+                                                                 const uint8_t* message,
+                                                                 size_t message_length,
+                                                                 const uint8_t* user_context,
+                                                                 size_t user_context_length,
+                                                                 uint8_t* auth_token,
+                                                                 size_t* auth_token_length,
+                                                                 uint8_t* encrypted_message,
+                                                                 size_t* encrypted_message_length)
+{
+    themis_status_t res = THEMIS_FAIL;
+    uint8_t iv[THEMIS_AUTH_SYM_IV_LENGTH] = {0};
+    uint8_t salt[THEMIS_AUTH_SYM_PBKDF2_SALT_LENGTH] = {0};
+    uint8_t auth_tag[THEMIS_AUTH_SYM_AUTH_TAG_LENGTH] = {0};
+    uint8_t derived_key[THEMIS_AUTH_SYM_KEY_LENGTH / 8] = {0};
+    struct themis_scell_auth_token_passphrase hdr = {0};
+    struct themis_scell_pbkdf2_context kdf = {0};
+    size_t auth_tag_length = 0;
+
+    /* Message length is currently stored as 32-bit integer, sorry */
+    if (message_length > UINT32_MAX) {
+        return THEMIS_INVALID_PARAMETER;
+    }
+
+    kdf.iteration_count = THEMIS_AUTH_SYM_PBKDF2_ITERATIONS;
+    kdf.salt = salt;
+    kdf.salt_length = sizeof(salt);
+
+    hdr.alg = THEMIS_AUTH_SYM_PASSPHRASE_ALG;
+    hdr.iv = iv;
+    hdr.iv_length = sizeof(iv);
+    hdr.auth_tag = auth_tag;
+    hdr.auth_tag_length = sizeof(auth_tag);
+    hdr.message_length = (uint32_t)message_length;
+    /* KDF pointer is ignored but size is important */
+    hdr.kdf_context = NULL;
+    hdr.kdf_context_length = (uint32_t)themis_scell_pbkdf2_context_size(&kdf);
```

Comment:
here we hardcode pbkdf2 as kdf context. would be better to add some general function like `themis_scell_context_size` which will have `switch` statement with all supported algorithms and default case which use algorithm `THEMIS_AUTH_SYM_PASSPHRASE_ALG & SOTER_SYM_KDF_MASK`. And each case call specific function for chosen algo.

Reply:
I figured that if we are going to support only one default KDF for creating new messages then it would be okay to hardcode it in the encryption path since that's the only KDF that we will be using. I've thought (a little) about generalizing the code, but did not find any obvious clean solution and left it as is. However, we'd definitely need a general parser for decryption path, because when we introduce another KDF we still have to support both PBKDF2 and whatever we add next.

I'll give it a look before merging this PR. Hopefully some solution will emerge.

Reply:
@Lagovas, I've generalized the decryption code path, now it's less coupled with the KDF choice. Extensibility is important here because we may add new KDF-related code for decryption to handle new formats, but we still have to handle old formats too.

With current code adding a new KDF requires adding a branch to the switch in `themis_auth_sym_derive_decryption_key()` which calls some new helper like `themis_auth_sym_derive_decryption_key_argon2()`.

I've looked at the encryption code path but no insights yet. I think it is acceptable to hardcode PBKDF2 choice during decryption since it is not configurable. However, it would be nice to decouple it in a similar way. Stand by for updates, I'll give it a try later.

Reply:
@Lagovas, I've finally came up with generalized code for the encryption path as well. I've managed to avoid adding any constants like 'max KDF context size' which is pretty neat. Adding new KDF for encryption is now the same as for decryption: write `themis_auth_sym_derive_encryption_key_argon2()` and add it to the KDF switch in `themis_auth_sym_derive_encryption_key()`.

With this, Secure Cell encryption/decryption no longer hardcodes KDF-related calls. They are dispatched based on the algorithm ID. If we update the algorithm ID but forget to add KDF implementation for it then the tests will fail.

Reply:
now it looks much better)